feat(nextjs,backend,integration): Introduce dynamic keys from `clerkMiddleware` (#3525)
clerk/javascript: .changeset/young-pigs-live.md

---

> '@clerk/backend': minor
> '@clerk/nextjs': minor
> '@clerk/shared': minor

Reviewer: `@clerk/shared` looks untouched 👀 

```suggestion
'@clerk/backend': minor
'@clerk/nextjs': minor
```

---

> Introduces dynamic keys from `clerkMiddleware`, allowing access by server-side helpers like `auth`. Keys such as `signUpUrl`, `signInUrl`, `publishableKey` and `secretKey` are securely encrypted using AES algorithm.

Reviewer: ```suggestion
Introduces dynamic keys from `clerkMiddleware`, allowing access by server-side helpers like `auth`. Options such as `signUpUrl`, `signInUrl`, `publishableKey` and `secretKey` are securely encrypted using AES algorithm and made available to the Next.js backend.
```

Author: I was getting confused between "Keys" and "Options", nice feedback!

- `auth` couldn't assert the session token signature due to not having access to `secretKey` passed as an option.
- `auth.redirectToSignIn` doesn't redirect to `signInUrl` provided as option.

`CLERK_ENCRYPTION_KEY` must be defined when providing `secretKey` as an option, in order to securely propagated this value

> between servers. For more information regarding options propagation, refer to (TODO - Add docs URL here)

Author: I'll spin up the docs PRs to merge alongside this feature.

> - `clerkClient` from `@clerk/nextjs` should now be called as a function, and it's singleton is deprecated, in favor of reading dynamic keys from Next.js runtime.

Reviewer: ```suggestion
- `clerkClient` from `@clerk/nextjs` should now be called as a function, and its singleton form is deprecated. This change allows the Clerk backend client to read keys from the current request, which is necessary to support dynamic keys.
```

Author: Nice one on providing context regarding the change!
